:penguin: Fix Some Lint Issues for Debian-based Images

sdwilsh opened the pull request:
**What this PR does / why we need it**:
This fixes [DL3027](https://github.com/hadolint/hadolint/wiki/DL3027).  That surfaces [DL3088](https://github.com/hadolint/hadolint/wiki/DL3008), which this change adds to the ignore file for now.

Along the way, it also fixes [DL3009](https://github.com/hadolint/hadolint/wiki/DL3009) and partially fixes [DL3059](https://github.com/hadolint/hadolint/wiki/DL3059) for the Debian-based images.

sdwilsh commented:
This isn't ready yet.  It looks like we were depending on some of the recommended packages that are now not installed since one of the lint rules wants us to pass `--no-install-recommends`.

sdwilsh commented:
I'm not really sure what's going on with that Ubuntu failure, nor how to reproduce it.  Guidance would be appreciated!

mudler commented:
hey @sdwilsh thanks, this is looking really nice!

at a glance doesn't look related, but on the other hand never saw it before, so it doesn't look a typical flaky test, I'll try retriggering and see if it's consistent.

c0ffee commented on images/Dockerfile.ubuntu:
```diff
@@ -38,14 +40,15 @@ RUN apt install -y \
-    && apt-get clean
+    systemd-timesyncd \
+    xz-utils \
+  && apt-get clean \
```
the clean is not needed anymore: https://github.com/moby/moby/blob/03e2923e42446dbb830c654d0eec323a0b4ef02a/contrib/mkimage/debootstrap#L82-L105

sdwilsh replied:
Here and elsewhere: it also existed in two places.  The linter seems happy with it, so I've removed it.

c0ffee commented on images/Dockerfile.ubuntu-20-lts:
```diff
@@ -40,14 +42,15 @@ RUN apt install -y \
-    && apt-get clean
+    systemd-timesyncd \
+    xz-utils \
+  && apt-get clean \
```
the clean is not needed anymore: https://github.com/moby/moby/blob/03e2923e42446dbb830c654d0eec323a0b4ef02a/contrib/mkimage/debootstrap#L82-L105

c0ffee commented on images/Dockerfile.ubuntu-22-lts:
```diff
@@ -40,14 +42,15 @@ RUN apt install -y \
-    && apt-get clean
+    systemd-timesyncd \
+    xz-utils \
+  && apt-get clean \
```
the clean is not needed anymore: https://github.com/moby/moby/blob/03e2923e42446dbb830c654d0eec323a0b4ef02a/contrib/mkimage/debootstrap#L82-L105

c0ffee commented on images/Dockerfile.debian:
```diff
@@ -38,14 +37,15 @@ RUN apt install -y \
-    && apt-get clean
+    systemd-timesyncd \
+    xz-utils \
+  && apt-get clean \
```
the clean is not needed anymore: https://github.com/moby/moby/blob/03e2923e42446dbb830c654d0eec323a0b4ef02a/contrib/mkimage/debootstrap#L82-L105

mudler reviewed (dismissed):
Looking good!

sdwilsh commented:
Adding some additional rigor to this, since @c0ffee had asked if this was parity with what we had before.  The short answer is no.

I grabbed the list of packages that were installed on master (https://github.com/kairos-io/kairos/commit/9b34a0b24af66fbb613856674021d281e26a4c3d) and with this fix (https://github.com/kairos-io/kairos/pull/821/commits/9a9917729cf34bcfb693a4ff206bbded9f2b15b0), and dumped them in this gist: https://gist.github.com/sdwilsh/4730c01787bf285f19d2dad98d0471cd
Just looking at line count differences, we can see the set of packages is different.  The deltas are what will actually be useful to us humans, however.

# debian
## Missing
<details>
<summary>Expand for full list of 123 packages</summary>
<ul>
<li>binutils
<li>binutils-common
<li>binutils-x86-64-linux-gnu
<li>bsd-mailx
<li>busybox
<li>console-setup
<li>console-setup-linux
<li>cryptsetup
<li>cryptsetup-bin
<li>dirmngr
<li>dmeventd
<li>dmraid
<li>efibootmgr
<li>ethtool
<li>exim4-base
<li>exim4-config
<li>exim4-daemon-light
<li>firmware-linux-free
<li>fuse3
<li>gnupg
<li>gnupg-l10n
<li>gnupg-utils
<li>gnupg1-l10n
<li>gpg
<li>gpg-agent
<li>gpg-wks-client
<li>gpg-wks-server
<li>gpgconf
<li>gpgsm
<li>grub-efi-amd64-signed
<li>isc-dhcp-common
<li>kbd
<li>keyboard-configuration
<li>keyutils
<li>krb5-locales
<li>libaio1
<li>libassuan0
<li>libatm1
<li>libbinutils
<li>libctf-nobfd0
<li>libctf0
<li>libdevmapper-event1.02.1
<li>libdmraid1.0.0.rc16
<li>libevent-2.1-7
<li>libevent-core-2.1-7
<li>libglib2.0-data
<li>libgnutls-dane0
<li>libgpm2
<li>libgprofng0
<li>libice6
<li>libidn12
<li>libisns0
<li>libjansson4
<li>libksba8
<li>libldap-common
<li>liblocale-gettext-perl
<li>liblockfile-bin
<li>liblockfile1
<li>liblvm2cmd2.03
<li>libmpdec3
<li>libnfsidmap1
<li>libnftables1
<li>libnl-3-200
<li>libnl-genl-3-200
<li>libnpth0
<li>libnss-systemd
<li>libopeniscsiusr
<li>libpam-cap
<li>libpkgconf3
<li>libpython3-stdlib
<li>libpython3.11-minimal
<li>libpython3.11-stdlib
<li>libsasl2-modules
<li>libsm6
<li>libsqlite3-0
<li>libunbound8
<li>libx11-6
<li>libx11-data
<li>libxau6
<li>libxcb1
<li>libxdmcp6
<li>libxext6
<li>libxmu6
<li>libxmuu1
<li>libxt6
<li>lvm2
<li>mdadm
<li>media-types
<li>mokutil
<li>nbd-client
<li>ncurses-term
<li>nfs-common
<li>nftables
<li>open-iscsi
<li>os-prober
<li>pigz
<li>pinentry-curses
<li>pkg-config
<li>pkgconf
<li>pkgconf-bin
<li>psmisc
<li>publicsuffix
<li>python3
<li>python3-greenlet
<li>python3-minimal
<li>python3-msgpack
<li>python3-pynvim
<li>python3.11
<li>python3.11-minimal
<li>rpcbind
<li>shared-mime-info
<li>shim-helpers-amd64-signed
<li>shim-signed
<li>shim-signed-common
<li>shim-unsigned
<li>thin-provisioning-tools
<li>x11-common
<li>xauth
<li>xclip
<li>xdg-user-dirs
<li>xkb-data
<li>xxd
<li>zerofree
</details>

## Added
* xz-utils (I should just pull this out into a new PR; I only noticed it in the logs because it being missing was an error for Ubuntu)

# ubuntu
## Missing
<details>
<summary>Expand for the full list of 107 packages</summary
<ul>
<li>binutils
<li>binutils-common
<li>binutils-x86-64-linux-gnu
<li>busybox-initramfs
<li>bzip2
<li>console-setup
<li>console-setup-linux
<li>cryptsetup
<li>cryptsetup-bin
<li>cryptsetup-initramfs
<li>dmeventd
<li>efibootmgr
<li>ethtool
<li>finalrd
<li>firmware-sof-signed
<li>fuse3
<li>grub-efi-amd64-signed
<li>initramfs-tools-bin
<li>initramfs-tools-core
<li>isc-dhcp-common
<li>kbd
<li>keyboard-configuration
<li>keyutils
<li>klibc-utils
<li>krb5-locales
<li>libaio1
<li>libatm1
<li>libbinutils
<li>libctf-nobfd0
<li>libctf0
<li>libdbus-glib-1-2
<li>libdevmapper-event1.02.1
<li>libdpkg-perl
<li>libevdev2
<li>libevent-core-2.1-7a
<li>libfile-fcntllock-perl
<li>libgdbm-compat4
<li>libgdbm6
<li>libgprofng0
<li>libgudev-1.0-0
<li>libice6
<li>libisns0
<li>libklibc
<li>libldap-common
<li>liblocale-gettext-perl
<li>liblvm2cmd2.03
<li>libnfsidmap1
<li>libnl-3-200
<li>libnl-genl-3-200
<li>libnss-systemd
<li>libopeniscsiusr
<li>libpam-cap
<li>libperl5.34
<li>libplymouth5
<li>libsasl2-modules
<li>libsm6
<li>libupower-glib3
<li>libx11-6
<li>libx11-data
<li>libxau6
<li>libxcb1
<li>libxdmcp6
<li>libxext6
<li>libxmu6
<li>libxmuu1
<li>libxt6
<li>lvm2
<li>mdadm
<li>nbd-client
<li>ncurses-term
<li>networkd-dispatcher
<li>nfs-common
<li>open-iscsi
<li>os-prober
<li>packagekit-tools
<li>perl
<li>perl-modules-5.34
<li>pigz
<li>pkg-config
<li>plymouth
<li>plymouth-theme-ubuntu-text
<li>publicsuffix
<li>python3-distro-info
<li>python3-greenlet
<li>python3-msgpack
<li>python3-pynvim
<li>rpcbind
<li>sbsigntool
<li>secureboot-db
<li>shared-mime-info
<li>ssh-import-id
<li>systemd-hwe-hwdb
<li>systemd-resolved
<li>thermald
<li>thin-provisioning-tools
<li>unattended-upgrades
<li>upower
<li>wget
<li>x11-common
<li>xauth
<li>xclip
<li>xdg-user-dirs
<li>xkb-data
<li>xxd
<li>xz-utils
<li>zerofree
<li>zstd
</details>

## Added
* xz-utils (Odd, because the job certainly failed before I added it in https://github.com/kairos-io/kairos/pull/821/commits/d9e60b4c99ba45b4f8e5882cac4690f936229b29)

# ubuntu-20-lts
## Missing
<details>
<summary>Expand for the full list of 99 packages</summary>
<ul>
<li>binutils
<li>binutils-common
<li>binutils-x86-64-linux-gnu
<li>busybox-initramfs
<li>console-setup
<li>console-setup-linux
<li>cryptsetup
<li>cryptsetup-bin
<li>cryptsetup-initramfs
<li>cryptsetup-run
<li>dmeventd
<li>efibootmgr
<li>ethtool
<li>file
<li>finalrd
<li>fuse
<li>grub-efi-amd64-signed
<li>initramfs-tools-bin
<li>initramfs-tools-core
<li>isc-dhcp-common
<li>kbd
<li>keyboard-configuration
<li>keyutils
<li>klibc-utils
<li>krb5-locales
<li>libaio1
<li>libatm1
<li>libbinutils
<li>libctf-nobfd0
<li>libctf0
<li>libdbus-glib-1-2
<li>libdevmapper-event1.02.1
<li>libevdev2
<li>libfile-fcntllock-perl
<li>libgudev-1.0-0
<li>libice6
<li>libimobiledevice6
<li>libisns0
<li>libklibc
<li>liblocale-gettext-perl
<li>liblvm2cmd2.03
<li>libmagic-mgc
<li>libmagic1
<li>libnfsidmap2
<li>libnss-systemd
<li>libpam-cap
<li>libplist3
<li>libplymouth5
<li>libreadline5
<li>libsasl2-modules
<li>libsm6
<li>libtirpc-common
<li>libtirpc3
<li>libupower-glib3
<li>libusb-1.0-0
<li>libusbmuxd6
<li>libx11-6
<li>libx11-data
<li>libxau6
<li>libxcb1
<li>libxdmcp6
<li>libxext6
<li>libxmu6
<li>libxmuu1
<li>libxt6
<li>lvm2
<li>lz4
<li>mdadm
<li>nbd-client
<li>ncurses-term
<li>networkd-dispatcher
<li>nfs-common
<li>open-iscsi
<li>os-prober
<li>packagekit-tools
<li>plymouth
<li>plymouth-theme-ubuntu-text
<li>publicsuffix
<li>python3-distro
<li>python3-distro-info
<li>python3-greenlet
<li>python3-msgpack
<li>python3-neovim
<li>python3-pynvim
<li>rpcbind
<li>sbsigntool
<li>secureboot-db
<li>ssh-import-id
<li>thermald
<li>thin-provisioning-tools
<li>upower
<li>usbmuxd
<li>wget
<li>x11-common
<li>xauth
<li>xclip
<li>xkb-data
<li>xxd
<li>zerofree
</details>

## Added
<details>
<summary>Expand for the full list of 8 packages</summary>
<ul>
<li>readline-common
<li>software-properties-common
<li>systemd
<li>systemd-sysv
<li>systemd-timesyncd
<li>tzdata
<li>ucf
<li>xz-utils
</details>


# ubuntu-22-lts
## Missing
<details>
<summary>Expand for the full list of 103 packages</summary>
<ul>
<li>binutils
<li>binutils-common
<li>binutils-x86-64-linux-gnu
<li>busybox-initramfs
<li>bzip2
<li>console-setup
<li>console-setup-linux
<li>cryptsetup
<li>cryptsetup-bin
<li>cryptsetup-initramfs
<li>dmeventd
<li>efibootmgr
<li>ethtool
<li>finalrd
<li>firmware-sof-signed
<li>fuse3
<li>grub-efi-amd64-signed
<li>initramfs-tools-bin
<li>initramfs-tools-core
<li>isc-dhcp-common
<li>kbd
<li>keyboard-configuration
<li>keyutils
<li>klibc-utils
<li>libaio1
<li>libatm1
<li>libbinutils
<li>libctf-nobfd0
<li>libctf0
<li>libdbus-glib-1-2
<li>libdevmapper-event1.02.1
<li>libevdev2
<li>libevent-core-2.1-7
<li>libfile-fcntllock-perl
<li>libgudev-1.0-0
<li>libice6
<li>libimobiledevice6
<li>libisns0
<li>libklibc
<li>libldap-common
<li>liblocale-gettext-perl
<li>liblvm2cmd2.03
<li>libnfsidmap1
<li>libnl-3-200
<li>libnl-genl-3-200
<li>libnss-systemd
<li>libopeniscsiusr
<li>libpam-cap
<li>libplist3
<li>libplymouth5
<li>libsasl2-modules
<li>libsm6
<li>libupower-glib3
<li>libusb-1.0-0
<li>libusbmuxd6
<li>libx11-6
<li>libx11-data
<li>libxau6
<li>libxcb1
<li>libxdmcp6
<li>libxext6
<li>libxmu6
<li>libxmuu1
<li>libxt6
<li>lvm2
<li>mdadm
<li>nbd-client
<li>ncurses-term
<li>networkd-dispatcher
<li>nfs-common
<li>open-iscsi
<li>os-prober
<li>packagekit-tools
<li>pigz
<li>plymouth
<li>plymouth-theme-ubuntu-text
<li>publicsuffix
<li>python3-distro-info
<li>python3-greenlet
<li>python3-msgpack
<li>python3-neovim
<li>python3-pynvim
<li>rpcbind
<li>sbsigntool
<li>secureboot-db
<li>shared-mime-info
<li>ssh-import-id
<li>systemd-hwe-hwdb
<li>thermald
<li>thin-provisioning-tools
<li>unattended-upgrades
<li>upower
<li>usbmuxd
<li>wget
<li>x11-common
<li>xauth
<li>xclip
<li>xdg-user-dirs
<li>xkb-data
<li>xxd
<li>xz-utils
<li>zerofree
<li>zstdV
</details>

## Added
* xz-utils (Odd, because the job certainly failed before I added it in https://github.com/kairos-io/kairos/pull/821/commits/d9e60b4c99ba45b4f8e5882cac4690f936229b29)

sdwilsh commented:
I pulled out two of the safer changes that won't have merge conflicts into their own PRs.  We should talk about what we want to do about the missing packages before we merge this.  Some of those on this list we probably don't want to include, but some of them we might want to still!

sdwilsh commented:
`ubuntu-20-lts` is consistently failing on runs, so I do not think it's random.  The failure is this:
```
+run-qemu-datasource-tests *failed* | • [FAILED] [1258.333 seconds]
+run-qemu-datasource-tests *failed* | kairos autoinstall test [BeforeEach] reboots and passes functional tests has grubenv file [autoinstall-test]
+run-qemu-datasource-tests *failed* |   [BeforeEach] /test/tests/autoinstall_test.go:23
+run-qemu-datasource-tests *failed* |   [It] /test/tests/autoinstall_test.go:91
+run-qemu-datasource-tests *failed* |   [FAILED] Timed out after 1200.000s.
+run-qemu-datasource-tests *failed* |   Expected
+run-qemu-datasource-tests *failed* |       <string>: Could not chdir to home directory /home/kairos: No such file or directory\nping\n
+run-qemu-datasource-tests *failed* |   to equal
+run-qemu-datasource-tests *failed* |       <string>: ping\n
+run-qemu-datasource-tests *failed* |   In [BeforeEach] at: /go/pkg/mod/github.com/spectrocloud/peg@v0.0.0-20230207104055-5f48d834ddf6/matcher/helpers.go:173 @ 02/09/23 03:08:38.584
+run-qemu-datasource-tests *failed* | ------------------------------
+run-qemu-datasource-tests *failed* | SSSSSSSSSSSSSSSSSSSSSSSSSS
+run-qemu-datasource-tests *failed* | ------------------------------
+run-qemu-datasource-tests *failed* | [AfterSuite] 
+run-qemu-datasource-tests *failed* | /test/tests/tests_suite_test.go:38
+run-qemu-datasource-tests *failed* | Deleting VM ... /tmp/1362627273
+run-qemu-datasource-tests *failed* | Cleaning /tmp/1362627273
+run-qemu-datasource-tests *failed* | [AfterSuite] PASSED [0.033 seconds]
+run-qemu-datasource-tests *failed* | ------------------------------
+run-qemu-datasource-tests *failed* | Summarizing 1 Failure:
+run-qemu-datasource-tests *failed* |   [FAIL] kairos autoinstall test [BeforeEach] reboots and passes functional tests has grubenv file [autoinstall-test]
+run-qemu-datasource-tests *failed* |   /go/pkg/mod/github.com/spectrocloud/peg@v0.0.0-20230207104055-5f48d834ddf6/matcher/helpers.go:173
```

I'll try to look into this more tomorrow.

sdwilsh commented:
I'm having some issues trying to build the ISO so I can see what is going on locally for `ubuntu-20-lts`.  The build command I'm using is `./earthly.sh +iso --FLAVOR=ubuntu-20-lts`, but it fails trying to build the docs: https://ci.earthly.dev/logs?logId=65ab2a2b-5ebd-406f-9783-70b45927f93f

Am I using the right command to build this locally?  It seems like I shouldn't have a dependency problem because the project uses Earthly, but that kinda looks like what I'm hitting, alas.

mudler reviewed (approved):
Nice cleanup, thank you!
